sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcSchemaIOProvider.java

```diff
@@ -0,0 +1,172 @@
+    return "jdbc";
+  }
+
+  /**
+   * Returns the expected schema of the configuration object. Note this is distinct from the schema
+   * of the data source itself.
+   */
+  @Override
+  public Schema configurationSchema() {
+    return Schema.builder()
+        .addStringField("driverClassName")
+        .addStringField("jdbcUrl")
+        .addStringField("username")
+        .addStringField("password")
+        .addNullableField("connectionProperties", FieldType.STRING)
+        .addNullableField("connectionInitSqls", FieldType.iterable(FieldType.STRING))
+        .addNullableField("readQuery", FieldType.STRING)
+        .addNullableField("writeStatement", FieldType.STRING)
+        .addNullableField("fetchSize", FieldType.INT16)
+        .addNullableField("outputParallelization", FieldType.BOOLEAN)
+        .build();
+  }
+
+  /**
+   * Produce a SchemaIO given a String representing the data's location, the schema of the data that
+   * resides there, and some IO-specific configuration object.
+   */
+  @Override
+  public JdbcSchemaIO from(String location, Row configuration, Schema dataSchema) {
+    return new JdbcSchemaIO(location, configuration);
+  }
+
+  @Override
+  public boolean requiresDataSchema() {
+    return false;
+  }
+
+  @Override
+  public PCollection.IsBounded isBounded() {
+    return PCollection.IsBounded.BOUNDED;
+  }
+
+  /** An abstraction to create schema aware IOs. */
+  static class JdbcSchemaIO implements SchemaIO, Serializable {
+    protected final Row config;
+    protected final String location;
+
+    JdbcSchemaIO(String location, Row config) {
+      this.config = config;
+      this.location = location;
+    }
+
+    @Override
+    public Schema schema() {
+      return null;
+    }
+
+    @Override
+    public PTransform<PBegin, PCollection<Row>> buildReader() {
+      String readQuery;
+      if (config.getString("readQuery") != null) {
+        readQuery = config.getString("readQuery");
+      } else {
+        readQuery = String.format("SELECT f_int FROM %s", location);
+      }
+
+      JdbcIO.DataSourceConfiguration dataSourceConfiguration = getDataSourceConfiguration();
+
+      JdbcIO.ReadRows readRows =
+          JdbcIO.readRows()
+              .withDataSourceConfiguration(dataSourceConfiguration)
+              .withQuery(readQuery);
+
+      if (config.getInt16("fetchSize") != null) {
+        readRows = readRows.withFetchSize(config.getInt16("fetchSize"));
+      }
+      if (config.getBoolean("outputParallelization") != null) {
+        readRows = readRows.withOutputParallelization(config.getBoolean("outputParallelization"));
+      }
+      return readRows;
+    }
+
+    @Override
+    public PTransform<PCollection<Row>, PDone> buildWriter() {
+      String writeStatement;
+      if (config.getString("writeStatement") != null) {
+        writeStatement = config.getString("writeStatement");
+      } else {
+        writeStatement = String.format("INSERT INTO %s VALUES(?, ?, ?)", location);
```

Comment:
The number of ?s in this query should depend on the number of fields in the schema of the input PCollection. `(?, ?, ?)` will only work when writing exactly three fields. I'd suggest returning a PTransform that looks at the input PCollection in its expand method. Something like this:

```java
return new PTransform<PCollection<Row>, PDone>() {
  public PDone expand(PCollection<Row> input) {
    Schema schema = input.getSchema()
    // Generate writeStatement based on input schema
    
    return input.apply(JdbcIO.<Row>write()...);
  }
}
```

Reply:
Done.